zeroex/orderwatch/order_watcher.go

```diff
@@ -99,50 +93,321 @@ func New(meshDB *meshdb.MeshDB, blockWatcher *blockwatch.Watcher, orderValidator
 	return w, nil
 }
 
-// Start sets up the event & expiration watchers as well as the cleanup worker. Event
-// watching will require the blockwatch.Watcher to be started however.
-func (w *Watcher) Start() error {
-	w.setupMux.Lock()
-	defer w.setupMux.Unlock()
-	if w.isSetup {
-		return errors.New("Setup can only be called once")
+// Watch sets up the event & expiration watchers as well as the cleanup worker.
+// Event watching will require the blockwatch.Watcher to be started first. Watch
+// will block until there is a critical error or the given context is canceled.
+func (w *Watcher) Watch(ctx context.Context) error {
+	w.mu.Lock()
+	if w.wasStartedOnce {
+		w.mu.Unlock()
+		return errors.New("Can only start Watcher once per instance")
 	}
+	w.wasStartedOnce = true
+	w.mu.Unlock()
 
-	w.setupEventWatcher()
+	// Start the expiration watcher.
+	expirationErrChan := make(chan error, 1)
+	go func() {
+		expirationErrChan <- w.expirationWatcher.Watch(ctx, expirationPollingInterval)
+	}()
 
-	if err := w.setupExpirationWatcher(); err != nil {
+	// Start the main loop, which listens and responds to various events.
+	if err := w.mainLoop(ctx); err != nil {
 		return err
 	}
 
-	w.startCleanupWorker()
+	// If the main loop exits without an error, return the error that came from
+	// the expiration watcher (which might be nil).
+	return <-expirationErrChan
+}
 
-	w.isSetup = true
-	return nil
+func (w *Watcher) mainLoop(ctx context.Context) error {
+	// Set up the channel used for subscribing to block events.
+	blockEvents := make(chan []*blockwatch.Event, 10)
+	w.blockSubscription = w.blockWatcher.Subscribe(blockEvents)
+
+	cleanupTicker := time.NewTicker(cleanupInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			cleanupTicker.Stop()
+			w.blockSubscription.Unsubscribe()
+			close(blockEvents)
+			return nil
+		case expiredOrders := <-w.expirationWatcher.ExpiredItems():
+			w.handleExpiration(expiredOrders)
+		case <-cleanupTicker.C:
```

Comment:
@fabioberger one consequence of this change is that the cleanup worker runs _synchronously_ with all other handlers in `mainLoop`. In other words, while the cleanup worker is running, we will stop processing block events and expirations.

This new version has some nice benefits:

1. It allows us to handle errors synchronously and will make it much easier to implement graceful shutdowns in the future (see #161).
2. It helps prevent race conditions because no one else will be validating/updating/removing orders while the cleanup worker is running.
3. It may make it easier for us to implement rate-limiting for Etheruem RPC requests, but it's hard to say for sure until we actually implement it.

However, this does come with an obvious downside in terms of performance. You have a better sense of how long the cleanup worker will take to run and how it compares to the other tasks within the `orderwatch` package. Do you think this trade-off is worth it or do you want to go back to having the cleanup worker run asynchronously?

Reply:
The cleanup working needs to re-validate _every_ order stored in the DB. This will involve multiple Ethereum JSON RPC requests and so can take a considerable amount of time. I worry that this could take several seconds to minutes depending on the number of orders stored. With the number of orders in Mesh today, it probably takes a few seconds (which is not a deal-breaker), but I worry this approach won't work once the network has more orders (it would only take one stress-tester to make happen).

I feel like running it synchronously with all other handlers is not tenable. Let's discuss this some more offline.